go/libkb/upak_lite.go

```diff
@@ -0,0 +1,270 @@
+		return nil, err
+	}
+	return highChain.ToUPAKLite(user)
+}
+
+type HighSigChainLoader struct {
+	MetaContextified
+	user      *User
+	leaf      *MerkleUserLeaf
+	chain     *HighSigChain
+	chainType *ChainType
+	links     ChainLinks
+	ckf       ComputedKeyFamily
+	dirtyTail *MerkleTriple
+}
+
+type HighSigChain struct {
+	Contextified
+	uid        keybase1.UID
+	username   NormalizedUsername
+	chainLinks ChainLinks
+}
+
+func NewHighSigChainLoader(m MetaContext, user *User, leaf *MerkleUserLeaf) *HighSigChainLoader {
+	hsc := HighSigChain{
+		uid:      user.GetUID(),
+		username: user.GetNormalizedName(),
+	}
+	loader := HighSigChainLoader{
+		user:             user,
+		leaf:             leaf,
+		chain:            &hsc,
+		chainType:        PublicChain,
+		MetaContextified: NewMetaContextified(m),
+	}
+	loader.ckf.kf = user.GetKeyFamily()
+	return &loader
+}
+
+func (l *HighSigChainLoader) Load() (ret *HighSigChain, err error) {
+	// request new links (and the unverified tail) from the server
+	// and put them into the highSigChain
+	err = l.LoadFromServer()
+	if err != nil {
+		return nil, err
+	}
+	// verify the chain
+	err = l.chain.VerifyChain(l.M())
+	if err != nil {
+		return nil, err
+	}
+	// compute keys
+	err = l.VerifySigsAndComputeKeys()
+
+	return l.chain, nil
+}
+
+func (l *HighSigChainLoader) selfUID() (uid keybase1.UID) {
+	// for now let's always assume this isn't applicable, because there's
+	// no distinction for loading yourself
+	return uid
+}
+
+func (l *HighSigChainLoader) LoadFromServer() (err error) {
+	srv := l.GetMerkleTriple()
+	l.dirtyTail, err = l.chain.LoadFromServer(l.M(), srv, l.selfUID())
+	return err
+}
+
+func (hsc *HighSigChain) LoadFromServer(m MetaContext, t *MerkleTriple, selfUID keybase1.UID) (dirtyTail *MerkleTriple, err error) {
+	// get the high sigs from the server
+	// ------------------
+	m, tbs := m.WithTimeBuckets()
+
+	apiArg := APIArg{
+		Endpoint:    "sig/get_high",
+		SessionType: APISessionTypeREQUIRED,
+		Args:        HTTPArgs{"uid": S{Val: hsc.uid.String()}},
+		MetaContext: m,
+	}
+	resp, finisher, err := m.G().API.GetResp(apiArg)
+	if err != nil {
+		return nil, err
+	}
+	if finisher != nil {
+		defer finisher()
+	}
+	recordFin := tbs.Record("HighSigChain.LoadFromServer.ReadAll")
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		recordFin()
+		return nil, err
+	}
+	recordFin()
+
+	// parse the response
+	// ------------------
+	if val, err := jsonparser.GetInt(body, "status", "code"); err == nil {
+		if keybase1.StatusCode(val) == keybase1.StatusCode_SCDeleted {
+			return nil, UserDeletedError{}
+		}
+	}
+	var links ChainLinks
+	var lastLink *ChainLink
+
+	jsonparser.ArrayEach(body, func(value []byte, dataType jsonparser.ValueType, offset int, inErr error) {
+		var link *ChainLink
+
+		parentSigChain := &SigChain{} // because we don't want the cache to use these
+		link, err = ImportLinkFromServer(m.G(), parentSigChain, value, selfUID)
+		links = append(links, link)
```

Comment:
preallocating the array could be a perf boost, but could wait until profiling to do it

Reply:
+1